### DISCUSSION TITLE
[solved] tox3: list envs should include description by default

### DISCUSSION THREAD

**Discussion by ssbarnea** [tox-dev/tox#1826] [Ideas]
As tox4 is already counting as a breaking change, I would like to propose making `tox -a` and `tox -l` more like they do when `-v` is also present.

This is primary because the description feature is unknown to most tox users, i found even veterans being pleasantly surprised about discovering it.

I know that the reason why this was done like this was to keep the output machine parsable but I think we have two options that could help here:

* activate verbose mode when stdin isatty(), so when suing shell pipes it would still be quiet.
* assure that `tox -vq` is machine parsable and document it as workaround

The idea is to make the default output more informative to the users, by default. Verbosity is supposed to be used for debugging, not for essential information.

IMHO, replacing `with description if verbose` could be replaced with `without description if quiet` in order to improve options discoverability for users.

**Comment by gaborbernat** [comment]
TLDR. I don't think so. I'd like to keep us backward compatible on this. tox 4 want to be a breaking change for just a small margine of cases, things which are deprecated for years.

> I would like to propose making `tox -a` and `tox -l` more like they do when `-v` is also present.

``tox -a`` and ``tox -l`` are considered with tox 4 legacy and in maintenance mode (as is the entire legacy sub-command). Listing environments has now been moved into the list command (``tox4 list`` or the shortform ``tox4 l``). Which already works as you propose here:

```
  {run,r,run-parallel,p,depends,de,list,l,devenv,d,config,c,quickstart,q,legacy,le}
                              tox command to execute (default to legacy if not specified)
    run (r)                   run environments
    run-parallel (p)          run environments in parallel
    depends (de)              visualize tox environment dependencies
    list (l)                  list environments
    devenv (d)                sets up a development environment at ENVDIR based on the env's tox configuration specified
    config (c)                show tox configuration
    quickstart (q)            Command-line script to quickly tox config file for a Python project
    legacy (le)               legacy entry-point command
```

**Reply by gaborbernat** [reply]
FYI if you do not specify a sub-command we assume legacy for backward compatibility.

**Reply by ssbarnea** [reply]
I did not know about this but considering new information, maybe the only thing we need is adding a warning line (stderr) when deprecated options are used, like "Deprecated option -l used, use `list` instead to avoid future failures."

**Reply by gaborbernat** [reply]
That could still mess up the machine parsing that output, so I'm hesitant about it. I feel like we should just let legacy flags untouched for now and maybe start warning in like a year or something. Once everyone had plenty of time to migrate per our updated documentation, which is to be written.

**Reply by ssbarnea** [reply]
Out of curiosity, what are your feelings about adding `requires >= 4.0.0a3` to few projects I manage to enforce use of the newer release. What is the risk of breaking them? Do you see benefits on doing that for tox4 testing?

**Reply by gaborbernat** [reply]
It would definitely help to test. However, I can't make any commitments about breaking changes. I'm trying not to, but sometimes I miss things. So it's up to your discretion, but in general, it would help. We haven't tested it much yet tox 3 to tox 4 interaction, so maybe add it, but in a way, it's both tox 3 and tox 4 compatible, and then in the CI, just switch to tox 4 would be my recommendation but expect potentially bumps.
